awx/main/utils/inventory_vars.py

```diff
@@ -0,0 +1,259 @@
+        :return: None
+        """
+        logger.debug(f"InventoryVariable().update({value}, {invsrc_id}):")
+        # Move this source to the front of the queue by first deleting a
+        # possibly existing entry, and then add the new entry to the front.
+        self.delete(invsrc_id)
+        self._update_queue.append((invsrc_id, value))
+
+    def delete(self, invsrc_id: int) -> None:
+        """
+        Delete an inventory source from the variable.
+
+        :param int invsrc_id: The inventory source id.
+        :return: None
+        """
+        data_index = self._get_invsrc_index(invsrc_id)
+        # Remove last update from this source, if there was any.
+        if data_index is not None:
+            value = self._update_queue.pop(data_index)[1]
+            logger.debug(f"InventoryVariable().delete({invsrc_id}): {data_index=} {value=}")
+
+    def _get_invsrc_index(self, invsrc_id: int) -> int | None:
+        """Return the inventory source's position in the queue, or `None`."""
+        for i, entry in enumerate(self._update_queue):
+            if entry[0] == invsrc_id:
+                return i
+        return None
+
+    def _get_current_value(self) -> var_value:
+        """
+        Return the current value of the variable, or None if the variable has no
+        history.
+        """
+        return self._update_queue[-1][1] if self._update_queue else None
+
+    @property
+    def value(self) -> var_value:
+        """Read the current value of the variable."""
+        return self._get_current_value()
+
+    @property
+    def has_no_source(self) -> bool:
+        """True, if the variable is orphan, i.e. no source contains this var anymore."""
+        return not self._update_queue
+
+    def __str__(self):
+        """Return the string representation of the current value."""
+        return str(self.value or "")
+
+
+class InventoryGroupVariables(dict):
+    """
+    Represent all inventory variables from one group.
+
+    This dict contains all variables of a inventory group and their current
+    value under consideration of the inventory source update history.
+
+    Note that variables values cannot be `None`, use the empty string to
+    indicate that a variable holds no value. See also `InventoryVariable`.
+    """
+
+    def __init__(self, name: str = "") -> None:
+        """
+        :param str name: The name of the group, 'all' for the all-group.
+        :return: None
+        """
+        super().__init__()
+        self.name = name
+        # In _vars we keep all sources for a given variable. This enables us to
+        # find the current value for a variable, which is the value from the
+        # latest update which defined this variable.
+        self._vars: dict[str, InventoryVariable] = {}
+
+    def _sync_vars(self) -> None:
+        """
+        Copy the current values of all variables into the internal dict.
+
+        Call this everytime the `_vars` structure has been modified.
```

Comment:
This is a defensible practice when this is only used internally by other methods in this class. Looking around, this is pretty solidly an internal method, and I think the `_vars` is an internal data structure. Yeah, that seems to check out :+1: